pyquil/api/_qam.py

```diff
@@ -60,18 +61,27 @@ def load(self, executable):
         return self
 
     @_record_call
-    def write_memory(self, *, region_name: str, offset: int = 0, value=None):
+    def write_memory(self, *, region_name: str, offset: int = None,
+                     value: Optional[Union[int, float, Sequence[int], Sequence[float]]] = None):
         """
-        Writes a value into a memory region on the QAM at a specified offset.
+        Writes a value or chronologically unwraps a list of values into a memory region on
+        the QAM at a specified offset.
 
         :param region_name: Name of the declared memory region on the QAM.
         :param offset: Integer offset into the memory region to write to.
-        :param value: Value to store at the indicated location.
+        :param value: Value(s) to store at the indicated location.
         """
         assert self.status in ['loaded', 'done']
 
-        aref = ParameterAref(name=region_name, index=offset)
-        self._variables_shim[aref] = value
+        if isinstance(value, Sequence):
+            if offset != None:
+                warnings.warn("offset should be None when value is a Sequence")
```

Comment:
It's not entirely clear to me why the offset is precluded in this case. Is the idea is that a region of the form `DECLARE foo REAL[20]` should be loaded with `qam.write_memory('foo', value = <list of 20 things>)`, as opposed to several calls (e.g. 
```
qam.write_memory('foo', value = <list of 10 things>) 
qam.write_memory('foo', offset=10, value = <some other list of 10 things>) 
```
)?

Reply:
This is something that was [requested by @karalekas](https://github.com/rigetti/pyquil/pull/1079#discussion_r347106051) in the previous incarnation of this PR.

I also think that offset + list of values could be useful. Maybe the intention was to try and avoid the caller mistakenly passing an offset + list when they really wanted the old version of the API offset + single value?

In any case, if we add the bounds checking you mention above, then I don't personally see the harm in allowing offset + list of values.